drivers/dp/swdp_ll_pin_stm32.h

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2025 Google LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stm32_ll_gpio.h>
+#include "stm32_hsem.h"
+
+#define CPU_CLOCK CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC
+#define FAST_BITBANG_HW_SUPPORT 1
+
+static ALWAYS_INLINE void swdp_ll_pin_input(void *const base, uint8_t pin)
+{
+	GPIO_TypeDef *gpio = base;
+
+	z_stm32_hsem_lock(CFG_HW_GPIO_SEMID, HSEM_LOCK_DEFAULT_RETRY);
+
+	LL_GPIO_SetPinMode(gpio, BIT(pin), LL_GPIO_MODE_INPUT);
+
+	z_stm32_hsem_unlock(CFG_HW_GPIO_SEMID);
+}
+
+static ALWAYS_INLINE void swdp_ll_pin_output(void *const base, uint8_t pin)
+{
+	GPIO_TypeDef *gpio = base;
+
+	z_stm32_hsem_lock(CFG_HW_GPIO_SEMID, HSEM_LOCK_DEFAULT_RETRY);
+
+	LL_GPIO_SetPinMode(gpio, BIT(pin), LL_GPIO_MODE_OUTPUT);
+
+	z_stm32_hsem_unlock(CFG_HW_GPIO_SEMID);
+}
+
+static ALWAYS_INLINE void swdp_ll_pin_set(void *const base, uint8_t pin)
+{
+	GPIO_TypeDef *gpio = base;
+	uint32_t val;
+
+	z_stm32_hsem_lock(CFG_HW_GPIO_SEMID, HSEM_LOCK_DEFAULT_RETRY);
+
+	val = LL_GPIO_ReadOutputPort(gpio);
+	val |= BIT(pin);
+	LL_GPIO_WriteOutputPort(gpio, val);
+
+	z_stm32_hsem_unlock(CFG_HW_GPIO_SEMID);
```

Comment:
Is it significantly faster than just GPIO API? I guess z_stm32_hsem_lock/unlock will be optimized away on single core?

Reply:
Yes and yes, I think about 250khz vs 550khz on the C0, something like 800 on the H7 can't remember right now but yeah big difference. Lock and lock are indeed no-ops on single cores.